### ISSUE TITLE
WPTable HTML not parsed correctly in selenium

### ISSUE THREAD

**Issue by davidbenton** [stephenharris/WordPressBehatExtension#15]
I'm not sure exactly what's happening yet, but it seems like there's something in WPTableRow or WPTableCell that's incompatible with the HTML/DOM used for post lists, at least with JS on. In Goutte I get passing tests, but in Selenium 2, I get a failure for the same test.

As an example, here is the (lightly edited) output for one of "Then the post list table looks like" feature steps from Goutte:

```
      | Title            | Author       | Tags   | Date                |
      | Another Doc      | Another User |        | Published2017/01/04 |
      | Multilingual Doc | Another User |        | Published2017/01/03 |
      | English Doc      | Admin        |        | Published2017/01/02 |
```

The same post list from Selenium 2 gives me an error, "Unable to parse post list table. Found: Array". That array's contents in table node format would be:

```
      |  |
      | Another User |        | Published2017/01/04 |
      | Another User |        | Published2017/01/03 |
      | Admin        |        | Published2017/01/02 |
```

Note the missing header row labels and missing title column. I've verified that the WP admin ui is working and shows the expected header row and title column by visiting with a full browser.

**Comment by davidbenton** [comment]
I'm stumped about how to fix this, but I've got some additional debugging info. It seems like the missing column headings and the missing title column might be two issues.

The headings are missing because [this line](https://github.com/stephenharris/WordPressBehatExtension/blob/f996cacf9f149134bd791b5461016c63de44600a/src/Context/Page/Element/WPTableRow.php#L44) doesn't work as expected. Instead or returning as many cells as there are columns, it only returns 1 cell: the `td#cb` checkbox column head. If we replace the lines...

```
foreach ($this->findAll('css', 'td,th') as $i => $cellNode) { //cells can be th or td
    $this->cells[] = new WPTableCell($cellNode, $this->session);
};
```
...with the hack-y...
```
foreach ($this->findAll('css', 'td') as $i => $cellNode) {
    $this->cells[] = new WPTableCell($cellNode, $this->session);
};
foreach ($this->findAll('css', 'th') as $i => $cellNode) {
    $this->cells[] = new WPTableCell($cellNode, $this->session);
};
```
... the column headings are found and the Exception is no longer thrown. But we still have a malformed TableNode. The table now looks like:
```
      | Title        | Author | Tags                | Date |
      | Another User |        | Published2017/01/04 |      |
      | Another User |        | Published2017/01/03 |      |
      | Admin        |        | Published2017/01/02 |      |
```
(headings aren't correctly paired with columns)

**Comment by davidbenton** [comment]
Continuing from above, I noticed that the `visitingColumnIndex` property of [WPTableNodeVisitor](https://github.com/stephenharris/WordPressBehatExtension/blob/f996cacf9f149134bd791b5461016c63de44600a/src/Context/Page/Element/WPTableNodeVisitor.php#L79) seems to be the cause of the lost title column values. The first two columns both return 0 for `visitingColumnIndex`, causing `isCheckBoxColumn()` to drop the first column, as expected, but also the second.

If you pare this—from method `visitCell()`—down...
```
if (0 === $this->visitingRowIndex && $cell->hasClass('column-cb')) {
    $this->checkboxColumnIndex = $this->visitingColumnIndex;
}
//Skip column
if (! $this->isCheckBoxColumn($this->visitingColumnIndex)) {
    $this->tableNodeHash[$this->visitingRowIndex][] = $cell->getCleanedText();
}
```
...to just...
```
$this->tableNodeHash[$this->visitingRowIndex][] = $cell->getCleanedText();
```
...the title values come back, but we get an empty cell in the heading row:
```
      |                  | Title        | Author | Tags                | Date |
      | Another Doc      | Another User |        | Published2017/01/04 |      |
      | Multilingual Doc | Another User |        | Published2017/01/03 |      |
      | English Doc      | Admin        |        | Published2017/01/02 |      |
```

Any ideas about the root issue(s) or how to proceed @stephenharris?

**Comment by stephenharris** [comment]
It is assumed that in `WPTable::get_rows()` the first cell is the checkbox cell. Also, those cells are are `th` so in your fix you will have it has last.

In the header row all cells are `th` so this doesn't matter: the first column is the checkbox cell. But in content rows this means the first cell is the first `th` cell (I.e. title).

So it seems this is due to Selenium not coping with `td,th`(?). I will need to check if I've run any similar tests using that driver. 

Also I wonder if making assumptions that the checkbox column is the first row is a bit too fragile.

**Comment by stephenharris** [comment]
Minor correction to the above. In the header row, all but the checkbox cell are `th`s.  The result is still the same: columns in the correct order. This does mean simply reversing the order in your change will not fix the issue.

Also, this test passed: https://travis-ci.org/stephenharris/WordPressBehatExtension/jobs/207354301#L522

That was using Selenium 2.53.1 and Mink Selenium2 Driver 1.3.1. What versions are you using?

**Comment by stephenharris** [comment]
One thing that has just popped to mind is that the table is different on a small screen - it shows only the checkbox and the title columns. But I still wouldn't expect to see what you get. Worth bearing in mind though.

**Comment by davidbenton** [comment]
Yeah, I definitely wasn't proposing those changes as a solution! Just trying to give a little more info about what's happening.

I'm using Selenium server standalone 2.53.1 and ChromeDriver 2.27.440174 (webdriver).

Interesting about the window size. I'll look into that.

> Also I wonder if making assumptions that the checkbox column is the first row is a bit too fragile.

The code actually checks for a class and then sets `checkboxColumnIndex`.
```
if (0 === $this->visitingRowIndex && $cell->hasClass('column-cb')) {
    $this->checkboxColumnIndex = $this->visitingColumnIndex;
}
```
In some of my debugging, I saw `visitingColumnIndex` of 0 for more than one column, but I don't know why it wasn't being incremented, or was being reset.

**Comment by davidbenton** [comment]
This seems similar to https://github.com/minkphp/Mink/issues/340, but I am using Mink v1.7.1 which includes a fix for that issue, and I've tried using `findAll('xpath', 'td | th')` and get the same result (only the first selector matches).

**Comment by davidbenton** [comment]
I am not going to be able to debug this any more for now, but it seems likely that this is a dependency bug, and not a WordPressBehatExtension bug. I'll be working around this for now by verifying my "Then" condition in some other way for my `@javascript`scenarios.
